### PR TITLE
[JENKINS-38527] - Prevent NullPointerException in Slave#createLauncher() and add cause diagnostics

### PR DESCRIPTION
**Pull request by oleg-nenashev**
The original issue comes from the isUnix() unboxing, but we can also get into an issue later if we pass a null Channel instance to the logic.
This change adds some diagnostics which discovers potential root causes of such potential NPEs due to the race conditions with Computer reconnection

See [JENKINS-38527](https://issues.jenkins-ci.org/browse/JENKINS-38527) raised by @ndeloof (isUnix()).

No tests, because I am not sure how to reliably reproduce such issue in JTH. `this != computer.getNode()` also may happen only due to the race condition

### Changelog entries

Proposed changelog entries:

* Entry 1: Prevent NullPointerException in Slave#createLauncher() when creating command launchers for disconnected or reconnected agents.
* ...

### Submitter checklist

- [x] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/Slave.java [hunk 1]
```diff
@@ -412,13 +417,44 @@ public URL getURL() throws IOException {
...
+                String message = "Issue with creating launcher for agent " + name + ". Computer has been reconnected";
+                if (LOGGER.isLoggable(Level.WARNING)) {
+                    LOGGER.log(Level.WARNING, message, new IllegalStateException("Computer has been reconnected, this Node instance cannot be used anymore"));
+                }
+                return new Launcher.DummyLauncher(listener);
+            }
+            
+            // RemoteLauncher requires an Channel instance to operate correctly
+            final Channel channel = c.getChannel();
+            final Boolean isUnix = channel != null ? c.isUnix() : null; // No sense to check it if the channel is down
```
Comment (stephenc) [inline core/src/main/java/hudson/model/Slave.java]: `channel != null && !channel.isClosingOrClosed()`

**Review by stephenc — CHANGES_REQUESTED** [review]
I think you need to guard against a closing or closed channel in the check... (though if this is a `VirtualChannel` you cannot IIRC in which case this code is fine)

File: core/src/main/java/hudson/model/Slave.java (hunk 1 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Slave.java]: It should have failed later in such case, but I agree it worth having an explicit check

Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Slave.java]: I expected `channel.call()` to handle it, but it does not...
The only handler is https://github.com/jenkinsci/remoting/blob/a762e0703a04ae41fbd413b2379d57f80fc282b2/src/main/java/hudson/remoting/Channel.java#L604-L605 , which effectively checks for the closed channel after a bunch of operations.

IMHO `Channel` should provide a new API for user-space calls which should not be performed in parallel with close. Making it a default behavior seems to be a breaking change though, needs much testing

Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Slave.java]: Created https://issues.jenkins-ci.org/browse/JENKINS-45023

**Review by stephenc — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@reviewbybees done, passing to @jenkinsci/code-reviewers

File: core/src/main/java/hudson/model/Slave.java [hunk 2]
```diff
@@ -412,13 +417,61 @@ public URL getURL() throws IOException {
...
+                                         "It is an invalid channel state, please report a bug to Jenkins if you see it.", 
+                                         listener);
+                return new Launcher.DummyLauncher(listener);
+            }
+            
+            return new RemoteLauncher(listener, channel, isUnix).decorateFor(this);
+        }
+    }
+    
+    private void reportLauncerCreateError(@Nonnull String humanReadableMsg, @CheckForNull String exceptionDetails, @Nonnull TaskListener listener) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Slave.java]: Typo.

**Review by jglick — COMMENTED** [review]
Seems to be updated in part by https://github.com/jenkinsci/remoting/pull/175.

File: core/src/main/java/hudson/model/Slave.java (hunk 2 above)
Comment (daniel-beck) [reply core/src/main/java/hudson/model/Slave.java]: Luckily `private` :-)

**Pull request merged**